### PR TITLE
issue: 3014055 Replace select(2) in the prepare phase

### PR DESCRIPTION
**Pull request by pasis**
If connect(2) returns a recoverable error, we try to poll for an event
and check error on the socket with getsockopt(SO_ERROR). Originally this
was done using select(2) which doesn't support file descriptor value
greater than FD_SETSIZE which is 1024.

A socket fd greater or equal to 1024 makes select(2) fail and this fail
isn't handled. In turn, this makes getsockopt(SO_ERROR) return
EINPROGRESS since connection isn't established yet and sockperf
terminates with the message "Can`t connect socket".

As solution for this issue and limitation, replace select(2) with
poll(2). Similarly to original behavior, don't terminate on polling
errors. Terminate only if getsockopt() reports an error.

### CODE REVIEW

File: src/client.cpp [hunk 1]
```diff
@@ -819,22 +819,14 @@ static int _connect_check_vma(int ifd) {
...
-
-    FD_SET(ifd, &wfds);
-    FD_SET(ifd, &rfds);
-    if (ifd > max_fd) max_fd = ifd;
-
-    select(max_fd + 1, &rfds, &wfds, NULL, &tv);
-    if (FD_ISSET(ifd, &wfds) || FD_ISSET(ifd, &rfds)) {
+    struct pollfd fds = { .fd = ifd, .events = POLLIN | POLLOUT, };
+#define POLL_TIMEOUT_MS 1
+    int ret = poll(&fds, 1, POLL_TIMEOUT_MS);
```
Comment (igor-ivanov) [inline src/client.cpp]: please do it under `#ifndef WIN32` and keep `select()` approach as temporary solution for windows

Comment (pasis) [reply src/client.cpp]: done.

File: src/client.cpp [hunk 2]
```diff
@@ -819,22 +819,32 @@ static int _connect_check_vma(int ifd) {
...
+    bool avail;
+
+#define POLL_TIMEOUT_MS 1
+
+#ifdef WIN32
     fd_set rfds, wfds;
-    struct timeval tv;
-    tv.tv_sec = 0;
-    tv.tv_usec = 500;
+    struct timeval tv = { .tv_sec = 0, .tv_usec = POLL_TIMEOUT_MS * 1000, };
```
Comment (igor-ivanov) [inline src/client.cpp:829]: May be 500 instead of 1000 to be aligned with previous.

**Review by igor-ivanov — APPROVED** [review]

File: src/client.cpp (hunk 2 above)
Comment (pasis) [reply src/client.cpp:829]: `poll(2)` API uses milliseconds, so the lowest value is 1ms. For select(2) I would like to use the same timeout to have a single point of knowledge for these timeouts.

Also a timeout is an unlikely event in this scenario.

**Pull request merged**